oem/ibm/libpldmresponder/bios_oem_ibm.cpp

```diff
@@ -0,0 +1,156 @@
+                             .getDbusProperty<std::vector<std::string>>(
+                                 objectPath.c_str(), namesProperty,
+                                 ibmCompatible[0].c_str());
+            writeFile(value[0]);
+            return value[0];
+        }
+        catch (const sdbusplus::exception_t& e)
+        {
+            error(
+                " Error getting Names property, ERROR={ERROR} PATH={PATH} INTERFACE={INTERFACE}",
+                "ERROR", e.what(), "PATH", searchpath, "INTERFACE",
+                ibmCompatible[0]);
+        }
+    }
+
+    return readFile();
+}
+
+/** @brief callback function invoked when interfaces get added from
+ *      Entity manager
+ *
+ *  @param[in] msg - Data associated with subscribed signal
+ */
+void pldm::responder::oem::ibm::bios::Handler::ibmCompatibleAddedCallback(
+    sdbusplus::message::message& msg)
+{
+    sdbusplus::message::object_path path;
+
+    std::map<std::string,
+             std::map<std::string,
+                      std::variant<std::string, std::vector<std::string>>>>
+        interfaceMap;
+
+    msg.read(path, interfaceMap);
+
+    if (!interfaceMap.contains(compatibleInterface))
+    {
+        return;
+    }
+    // Get the "Name" property value of the
+    // "xyz.openbmc_project.Configuration.IBMCompatibleSystem" interface
+    const auto& properties = interfaceMap.at(compatibleInterface);
+
+    if (!properties.contains(namesProperty))
+    {
+        return;
+    }
+    auto names =
+        std::get<pldm::utils::Interfaces>(properties.at(namesProperty));
+
+    // get only the first system type
+    if (!names.empty())
+    {
+        systemType = names.front();
+        writeFile(systemType);
+    }
+
+    if (!systemType.empty())
+    {
+        ibmCompatibleMatchConfig.reset();
+    }
+}
+
+void pldm::responder::oem::ibm::bios::Handler::writeFile(std::string data)
+{
+    fs::create_directories(systemTypeFilePath.parent_path());
+    std::ofstream writeFile(systemTypeFilePath.string(),
+                            std::ios::out | std::ios::binary);
+    if (writeFile)
+    {
+        writeFile << data << std::endl;
+        info("SystemType written in the file : {SYSTEM_TYPE}", "SYSTEM_TYPE",
+             data);
+        writeFile.close();
+    }
+    else
+    {
+        error("Error in writting SystemType in the file : {SYSTEM_TYPE}",
+              "SYSTEM_TYPE", data);
+    }
+}
+
+std::optional<std::string> pldm::responder::oem::ibm::bios::Handler::readFile()
+{
+    std::string data;
+    std::ifstream readFile(systemTypeFilePath.string(),
+                           std::ios::in | std::ios::binary);
+
+    if (readFile)
+    {
+        readFile >> data;
+        info("SystemType read from the file: {SYSTEM_TYPE}", "SYSTEM_TYPE",
+             data);
+        readFile.close();
+        return data;
+    }
+    else
+    {
+        error("Error in reading SystemType from the file");
+    }
+
+    return std::nullopt;
```

Comment:
how would the caller handle this in the case the return is a nullopt?

Reply:
Caller will read default bios attributes